File: go/reader/record_reader/multi_files_sequential_reader.go
package record_reader

import (
	"sync/atomic"

	"github.com/apache/arrow/go/v12/arrow"
	"github.com/apache/arrow/go/v12/arrow/array"
	"github.com/apache/arrow/go/v12/parquet/pqarrow"
	"github.com/milvus-io/milvus-storage/go/common/arrow_util"
	"github.com/milvus-io/milvus-storage/go/file/fragment"
	"github.com/milvus-io/milvus-storage/go/io/fs"
	"github.com/milvus-io/milvus-storage/go/storage/options/option"
)

type MultiFilesSequentialReader struct {
	fs         fs.Fs
	schema     *arrow.Schema
	files      []string
	nextPos    int
	options    *option.ReadOptions
	currReader array.RecordReader
	err        error
	ref        int64
}

func (m *MultiFilesSequentialReader) Retain() {
	atomic.AddInt64(&m.ref, 1)
}

func (m *MultiFilesSequentialReader) Release() {
	if atomic.AddInt64(&m.ref, -1) == 0 {
		if m.currReader != nil {
			m.currReader.Release()
			m.currReader = nil
		}
	}
}

func (m *MultiFilesSequentialReader) Schema() *arrow.Schema {
	return m.schema
}

func (m *MultiFilesSequentialReader) Next() bool {
	for true {
		if m.currReader == nil {
			if m.nextPos >= len(m.files) {
				return false
			}

			m.nextReader()
			if m.err != nil {
				return false
			}
			m.nextPos++
		}
		if m.currReader.Next() {
			return true
		}
		if m.currReader.Err() != nil {
			m.err = m.currReader.Err()
			return false
		}
		if m.currReader != nil {
			m.currReader.Release()
			m.currReader = nil
		}
	}
	return false
}

func (m *MultiFilesSequentialReader) Record() arrow.Record {
	if m.currReader != nil {
		return m.currReader.Record()
	}
	return nil
}

func (m *MultiFilesSequentialReader) Err() error {
	return m.err
}

func (m *MultiFilesSequentialReader) nextReader() {
	var fileReader *pqarrow.FileReader
	fileReader, m.err = arrow_util.MakeArrowFileReader(m.fs, m.files[m.nextPos])
	if m.err != nil {
		return
	}
	m.currReader, m.err = arrow_util.MakeArrowRecordReader(fileReader, m.options)
	return
}

func NewMultiFilesSequentialReader(fs fs.Fs, fragments fragment.FragmentVector, schema *arrow.Schema, options *option.ReadOptions) *MultiFilesSequentialReader {
	files := make([]string, 0, len(fragments))
	for _, f := range fragments {
		files = append(files, f.Files()...)
	}

	return &MultiFilesSequentialReader{
		fs:      fs,
		schema:  schema,
		options: options,
		files:   files,
		nextPos: 0,
		ref:     1,
	}
}

File: cpp/src/reader/multi_files_sequential_reader.cpp


#include "milvus-storage/reader/multi_files_sequential_reader.h"
#include <arrow/record_batch.h>
#include <arrow/type.h>
#include "milvus-storage/common/arrow_util.h"

namespace milvus_storage {

MultiFilesSequentialReader::MultiFilesSequentialReader(arrow::fs::FileSystem& fs,
                                                       const FragmentVector& fragments,
                                                       std::shared_ptr<arrow::Schema> schema,
                                                       const SchemaOptions& schema_options,
                                                       const ReadOptions& options)
    : fs_(fs), schema_(std::move(schema)), schema_options_(schema_options), options_(options) {
  for (const auto& fragment : fragments) {
    files_.insert(files_.end(), fragment.files().begin(), fragment.files().end());
  }
}

std::shared_ptr<arrow::Schema> MultiFilesSequentialReader::schema() const { return schema_; }

arrow::Status MultiFilesSequentialReader::ReadNext(std::shared_ptr<arrow::RecordBatch>* batch) {
  while (true) {
    if (!curr_reader_) {
      if (next_pos_ >= files_.size()) {
        batch = nullptr;
        return arrow::Status::OK();
      }

      auto s = MakeArrowFileReader(fs_, files_[next_pos_++]);
      if (!s.ok()) {
        return arrow::Status::UnknownError(s.status().ToString());
      }
      holding_file_reader_ = std::move(s.value());

      auto s2 = MakeArrowRecordBatchReader(*holding_file_reader_, schema_, schema_options_, options_);
      if (!s2.ok()) {
        return arrow::Status::UnknownError(s2.status().ToString());
      }
      curr_reader_ = std::move(s2.value());
    }

    std::shared_ptr<arrow::RecordBatch> tmp_batch;
    auto s = curr_reader_->ReadNext(&tmp_batch);
    if (!s.ok()) {
      return s;
    }

    if (tmp_batch == nullptr) {
      curr_reader_ = nullptr;
      holding_file_reader_ = nullptr;
      continue;
    }

    *batch = tmp_batch;
    return arrow::Status::OK();
  }
}

}  // namespace milvus_storage
